lib/ansible/modules/network/fortios/fortios_facts.py

```diff
@@ -0,0 +1,291 @@
+# Copyright 2019 Fortinet, Inc.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'status': ['preview'],
+                    'supported_by': 'community',
+                    'metadata_version': '1.1'}
+
+DOCUMENTATION = '''
+---
+module: fortios_facts
+version_added: "2.9"
+short_description: Get facts about fortios devices.
+description:
+    - Collects facts from network devices running the fortios operating
+      system. This module places the facts gathered in the fact tree keyed by the
+      respective resource name.  The facts module will always collect a
+      base set of facts from the device and can enable or disable
+      collection of additional facts.
+author:
+    - Miguel Angel Munoz (@mamunozgonzalez)
+    - Nicolas Thomas (@thomnico)
+    - Don Yao (@fortinetps)
+notes:
+    - Requires fortiosapi library developed by Fortinet
+    - Run as a local_action in your playbook
+requirements:
+    - fortiosapi>=0.9.8
+options:
+    host:
+        description:
+            - FortiOS or FortiGate IP address.
+        type: str
+        required: true
+    username:
+        description:
+            - FortiOS or FortiGate username.
+        type: str
+        required: true
+    password:
+        description:
+            - FortiOS or FortiGate password.
+        type: str
+        default: ""
+        required: true
+    vdom:
+        description:
+            - Virtual domain, among those defined previously. A vdom is a
+              virtual instance of the FortiGate that can be configured and
+              used as a different unit.
+        type: str
+        default: root
+        required: false
+    https:
+        description:
+            - Indicates if the requests towards FortiGate must use HTTPS protocol.
+        type: bool
+        default: true
+        required: false
+    ssl_verify:
+        description:
+            - Ensures FortiGate certificate must be verified by a proper CA.
+        type: bool
+        default: false
+        required: false
+    gather_subset:
+        description:
+            - When supplied, this argument will restrict the facts collected
+              to a given subset.  Possible values for this argument include
+              all, hardware, config, and interfaces.  Can specify a list of
```

Comment:
Please update the description based on what values are currently supported for `gather_subset`.
The pattern we've been following of {{ network_os }}_facts modules is, apart from the valid choices, it also supports - `all` and `!{{ fact_to_collect }}`.

With `gather_subset` set to `all`, the module will collect all the facts in `VALID_SUBSETS` and if there's a fact prepended with `!` specified, that will not be collected.

Reply:
Hi Nilashish,
I understand that the standard implementation with "all" and "!" options. But not all the network/firewall devices are design/build the same way. The problem for FortiOS is, the "all" option would 1) generate tons of information (to be exact, as I count, 410 different facts) and gather them all will take some time, and impact the performance, 2) none of our customers would need this huge amount of info at once, none of them want facts gathering impact their production performance, most of them just need one specify query for a specific set of facts, 3) some facts gathering generate binary data which need to log/write into local file, and some additional arguments are required. Also we don't want to support "!' at this time, since none of our customers need this. Would you please take this into consideration, this first implementation just for handful of facts, which are most important (3 facts out of 410 in total) for one big customer, and they want to get this in Ansible 2.9 so they can deploy in their production environment.

Reply:
@fortinetp Thanks for the clarification. Having `all` and `!` seems like a good idea to me, in the sense, that if let's say the module supports gathering 10 facts and user wants to collect 9 of them, except 1, then in that case the user can just mention `all` and `!{{ fact_to_not_collect }}` as opposed to specifying 9 facts. As you can understand, this would scale well with a large number of facts.

However, as you specified, your customers wouldn't be needing so many facts at one time, we can hold this thought for future releases and add it if deemed necessary.